### PR TITLE
Add Exception telemetry, upgrade telemetry object to ITelemetryService2

### LINKED ISSUE #246: Add telemetry with the exception call stack

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #246.

This adds a TraceListener which will respond to the events that StreamJsonRpc sends through its TraceSource. It checks for the specific trace event ID that the RPC library uses, and checks to make sure the arguments are sane. Adding more listeners or changing the trace level should be safe and won't break the listener.

Property naming is up to us, since it's a new event, but I mirrored some of what I saw in the extension, that is, camelCase names. The event name is the same as what the extension is using for its language server telemetry, but with a `/` afterward so we don't have any name conflicts and can do our own thing. I picked `rpc.exception` for this event, but it can be changed if wanted.

### CODE REVIEW

File: src/LanguageServer/Impl/LanguageServer.cs [hunk 1]
```diff
@@ -14,17 +14,17 @@
...
-using System.Linq;
-using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.Python.LanguageServer.Services;
 using Microsoft.PythonTools.Analysis;
 using Microsoft.PythonTools.Analysis.Infrastructure;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 using StreamJsonRpc;
+using System;
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/LanguageServer.cs]: Reverse sorting (set it in VS Tools | Options | Text Editor | C# | Advanced)

File: src/LanguageServer/Impl/TelemetryRpcTraceListener.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            }
+
+            // See JsonRpc.StripExceptionToInnerException.
+            if (exception is TargetInvocationException || (exception is AggregateException && exception.InnerException != null)) {
+                exception = exception.InnerException;
+            }
+
+            // Exceptions we expect to throw and pass back over RPC, and therefore should not record.
+            switch (exception) {
+                case InvalidOperationException _:
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/TelemetryRpcTraceListener.cs]: In `Rename` I'd change `InvalidOperationException` to `RenameException: NonFatalLanguageServerException` where `NonFatalLanguageServerException: LanguageServerException` and then filter out `NonFatalLanguageServerException`. This way actual bugs that cause `InvalidOperationException` will be logged.

File: src/LanguageServer/Impl/LanguageServer.cs (hunk 1 above)
Comment (jakebailey) [reply src/LanguageServer/Impl/LanguageServer.cs]: Bah, VS reset my settings again...

File: src/LanguageServer/Impl/TelemetryRpcTraceListener.cs (hunk 2 above)
Comment (jakebailey) [reply src/LanguageServer/Impl/TelemetryRpcTraceListener.cs]: I've introduced a new exception and replaced Rename's use of InvalidOperationException with it.

**Pull request merged**